### ISSUE TITLE
`serve` run one more worker

### ISSUE THREAD

**Issue by vjik** [yiisoft/yii-console#213]
For example, `serve -w2` runs 3 workers.

But I could be wrong and runned one parent process and two workers, that process requests.

**Comment by klsoft-web** [comment]
> For example, `serve -w2` runs 3 workers.
> 
> But I could be wrong and runned one parent process and two workers, that process requests.

Yes, I can confirm that, but it is not related to the `yiisoft/yii-console` package, since the following command produces the same result:
`PHP_CLI_SERVER_WORKERS=2 php -S localhost:8080 -t ./public`
So the issue should be closed.

**Comment by vjik** [comment]
It seems that `PHP_CLI_SERVER_WORKERS` specifies the number of forks created by the PHP CLI server. For example, when `PHP_CLI_SERVER_WORKERS` is set to `2`, PHP creates two child processes, resulting in three workers in total: one master process and two forked processes.

**Comment by samdark** [comment]
That is expected.

**Comment by vjik** [comment]
> That is expected.

Option description is "Workers number the server will start with". Why running 3 workers with "2" value is expected?

**Comment by klsoft-web** [comment]
> Option description is "Workers number the server will start with". Why running 3 workers with "2" value is expected?

What about replacing it with:
`The number of fork workers that the server will start with`

**Comment by vjik** [comment]
As a user of `serve` command, I’m interested in the actual number of workers. My suggestion is usage `value - 1` for `PHP_CLI_SERVER_WORKERS`.

**Comment by klsoft-web** [comment]
> As a user of `serve` command, I’m interested in the actual number of workers. My suggestion is usage `value - 1` for `PHP_CLI_SERVER_WORKERS`.

The [documentation on the PHP built-in web server](https://www.php.net/manual/en/features.commandline.webserver.php) does not make it clear what the purpose of the additional process is. 
Is it responsible for handling requests or dispatching other workers?

**Comment by vjik** [comment]
> The [documentation on the PHP built-in web server](https://www.php.net/manual/en/features.commandline.webserver.php) does not make it clear what the purpose of the additional process is.
> Is it responsible for handling requests or dispatching other workers?

It should be checked.

**Comment by klsoft-web** [comment]
> It should be checked.

I have checked.
Add a lock condition to the `SessionMiddleware`.
```php
if (isset($queryParams['lock'])) {
  sleep(60*10);
}
```

A maximum of one request can be handled at once by one worker.
` ./yii serve --workers=1`

A maximum of three requests can be handled at once by two workers. 
` ./yii serve --workers=2`

However, I wait until three requests have been accepted in the log. 
```
127.0.0.1:39020 Accepted
127.0.0.1:59764 Accepted
127.0.0.1:38730 Accepted
```

**Comment by vjik** [comment]
So, `--workers=2` gives 3 workers. It's unexpected.

@samdark WDYT?

**Comment by samdark** [comment]
@vjik 3 workers or 3 processes?

**Comment by samdark** [comment]
If that's three workers, then that's not expected and should be fixed.

**Comment by klsoft-web** [comment]
> If that's three workers, then that's not expected and should be fixed.

Based on [this call](https://github.com/php/php-src/blob/100e8f353ea8ef841339ba4c74a0966a7eea5157/sapi/cli/php_cli_server.c#L2575C16-L2575C20), the worker is a child process.
So there is one parent process and two child processes.
I also assume that when you set PHP_CLI_SERVER_WORKERS to 1, it is not a worker process because it is not a child process.

Therefore, we can create two workers or more, but not one.

**Comment by vjik** [comment]
> If that's three workers, then that's not expected and should be fixed.

Based on [research](https://github.com/yiisoft/yii-console/issues/213#issuecomment-5680477248) above, one parent process + two forks. But all three processes handle requests. So, by fact it is three workers.

**Comment by klsoft-web** [comment]
> So, by fact it is three workers.

More precisely, there are three processes. Two of these are workers, since there are only two child processes.

**Comment by samdark** [comment]
Okay, I've checked the source code. `PHP_CLI_SERVER_WORKERS` means how many forks are started **additionally to the main process**. So `PHP_CLI_SERVER_WORKERS` is the primary process + 2 forks, and **all three are processing requests**.

So it is a bit confusing, but Yii's serve is aligned with `PHP_CLI_SERVER_WORKERS` and I don't think we should change that. It might be worth mentioning in the command help, such as "start N workers in addition to the main process."
